nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      String consequent = getConsequent(clause, tree, analysis, state, callee);
+
+      // Find a single value constraint that is not already known. If more than one arguments with
+      // unknown
+      // nullness affects the method's result, then ignore this clause.
+      Node arg = null;
+      Nullness argAntecedentNullness = null;
+      boolean supported =
+          true; // Set to false if the rule is detected to be one we don't yet support
+
+      for (int i = 0; i < antecedent.length; ++i) {
+        String valueConstraint = antecedent[i].trim();
+        if (valueConstraint.equals("_")) {
+          continue;
+        } else if (valueConstraint.equals("false") || valueConstraint.equals("true")) {
+          // We handle boolean constraints in the case that the boolean argument is the result
+          // of a null or not-null check. For example,
+          // '@Contract("true -> true") boolean func(boolean v)'
+          // called with 'func(obj == null)'
+          // can be interpreted as equivalent to
+          // '@Contract("null -> true") boolean func(@Nullable Object v)'
+          // called with 'func(obj)'
+          // This path unwraps null reference equality and inequality checks
+          // to pass the target (in the above example, 'obj') as arg.
+          Node argument = node.getArgument(i);
+          // isNullTarget is the variable side of a null check. For example, both 'e == null'
+          // and 'null == e' would return the node representing 'e'.
+          Optional<Node> isNullTarget = argument.accept(NullEqualityVisitor.IS_NULL, inputs);
+          // notNullTarget is the variable side of a not-null check. For example, both 'e != null'
+          // and 'null != e' would return the node representing 'e'.
+          Optional<Node> notNullTarget = argument.accept(NullEqualityVisitor.NOT_NULL, inputs);
+          // It is possible for at most one of isNullTarget and notNullTarget to be present.
+          Node nullTestTarget = isNullTarget.orElse(notNullTarget.orElse(null));
+          if (nullTestTarget == null) {
             supported = false;
             break;
-          } else if (valueConstraint.equals("!null")
-              && inputs.valueOfSubNode(node.getArgument(i)).equals(Nullness.NONNULL)) {
-            // We already know this argument can't be null, so we can treat it as not part of the
-            // clause
-            // for the purpose of deciding the non-nullness of the other arguments.
+          }
+          // isNullTarget is equivalent to 'null ->' while notNullTarget is equivalent
+          // to '!null ->'. However, the valueConstraint may reverse the check.
+          boolean inverted = isNullTarget.isPresent() == valueConstraint.equals("false");
+          Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
+          Nullness targetNullness = inputs.valueOfSubNode(nullTestTarget);
```

Comment:
@carterkozak is this concern no longer present given the reduced scope of this PR?

Reply:
I think so, if I understand the concern, but the system is complex and we are in a way relying on the value here in a way: not to produce a static error, but rather to whittle down statically-known constraints until there's only one left that's not guaranteed true.

I could update the following line to `antecedentNullness.equals(targetNullness) && !antecedentNullness.equals(Nullness.NULL)` to prevent us from taking `x == null` as true where `final Object x = null` into account, while still matching behavior of the guaranteed-`!null` contract. What do you think?

Reply:
> I think so, if I understand the concern, but the system is complex and we are in a way relying on the value here in a way: not to produce a static error, but rather to whittle down statically-known constraints until there's only one left that's not guaranteed true.

Can you point me at the test case that takes advantage of this reasoning?  I think I understand, just want to make sure.

Reply:
Nevermind, I think it's the new test you added?  `nonNullBooleanDoubleContractTest()`?

Reply:
That's correct -- we're able to rule out the second parameter because it's guaranteed to satisfy the contract based on the object being non-null.

Reply:
Ok, so I think I get it.  The idea is that if we have an expression `e` of `@NonNull` type, and there is a check `e == null`, we know for sure that the check will evaluate to false.  So, if that check expression is the first argument to a function with contract `false, false -> fail`, we can prove that the failure case will occur iff the second argument is false.  So, continuing the example, if the second argument is `e2 == null` and `e2` is `@Nullable`, we can refine `e2` to `@NonNull` in the non-failure case.  Does that sound right?

Given this reasoning is localized to this function, I'm ok with keeping it as-is.  @lazaroclapp can give his opinion when he returns.

Reply:
That is exactly right

Reply:
I don't believe we track `Nullness.NULL` from fields, only `NULLABLE` -- I've added test coverage for local variables, which track explicit null assignment, but not fields as they wouldn't pass (and I don't think we want to verify that fields cannot produce `NULL` in the scope of this feature).